### PR TITLE
feat(sql): add min/max and regexp_replace functions for string values

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Note: `regexp_replace()` implementation doesn't support 4th argument (flags) and defaults to global case sensitive replacement.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 102 / 110 (92.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/InterpolationUtil.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunction.java|44|47|93.62%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunction.java|44|47|93.62%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunctionFactory.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 139 / 152 (91.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/InterpolationUtil.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java|30|35|85.71%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunction.java|43|46|93.48%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunction.java|43|46|93.48%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/MatchStrFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexUtils.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 167 / 184 (90.76%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/InterpolationUtil.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java|56|65|86.15%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunction.java|43|46|93.48%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunction.java|43|46|93.48%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ReplaceStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/MatchStrFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexUtils.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 167 / 184 (90.76%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/InterpolationUtil.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java|56|65|86.15%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunction.java|43|46|93.48%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunction.java|43|46|93.48%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ReplaceStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/MatchStrFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexUtils.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 173 / 190 (91.05%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/InterpolationUtil.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java|56|65|86.15%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunction.java|46|49|93.88%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunction.java|46|49|93.88%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ReplaceStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/MatchStrFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexUtils.java|6|6|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            if (cs == null) {
+                throw SqlException.$(replacementPos, "NULL replacement");
+            }
+            replacementStr = cs.toString();
+        }
+    }
+
+    // TODO(puzpuzpuz):
+    //  Get rid of this class in favor of StringSink once we drop support for Java 8 where j.u.r.Matcher
+    //  has method overloads for StringBuilder.
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java:173]: has **no** method overloads for StringBuilder? the word _no_ is missing?

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java:173]: Indeed, it's a typo. Will fix.

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java]: Done in f8022d255ec815f98350b81416b5b59d7bb7bf05

File: utils/dependency-reduced-pom.xml [hunk 2]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply utils/dependency-reduced-pom.xml]: Done in f8022d255ec815f98350b81416b5b59d7bb7bf05

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 173 / 190 (91.05%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/InterpolationUtil.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java|56|65|86.15%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunction.java|46|49|93.88%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunction.java|46|49|93.88%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ReplaceStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/MatchStrFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexUtils.java|6|6|100.00%|

**Review by jerrinot — APPROVED** [review]
+1. Also, I learned a new thing about keyed aggregations: At first, I thought: "Why the $#@! is are min/max aggregations so complicated. Just one sink should be enough." and then I realized it! :)

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 173 / 190 (91.05%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/InterpolationUtil.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexpReplaceStrFunctionFactory.java|56|65|86.15%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunction.java|46|49|93.88%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunction.java|46|49|93.88%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ReplaceStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MinStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/groupby/MaxStrGroupByFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/MatchStrFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/regex/RegexUtils.java|6|6|100.00%|

**Pull request merged**